shell/platform/android/io/flutter/embedding/engine/FlutterJNI.java

```diff
@@ -407,6 +411,41 @@ public void removeIsDisplayingFlutterUiListener(@NonNull FlutterUiDisplayListene
     flutterUiDisplayListeners.remove(listener);
   }
 
+  public static native void nativeImageHeaderCallback(
+      long imageGeneratorPointer, int width, int height);
+
+  /**
+   * Called by native as a fallback method of image decoding. There are other ways to decode images
+   * on lower API levels, they involve copying the native data _and_ do not support any additional
+   * formats, whereas ImageDecoder supports HEIF images. Unlike most other methods called from
+   * native, this method is expected to be called on a worker thread, since it only uses thread safe
+   * methods and may take multiple frames to complete.
+   */
+  @SuppressWarnings("unused")
+  @VisibleForTesting
+  @Nullable
+  public static Bitmap decodeImage(@NonNull ByteBuffer buffer, long imageGeneratorAddress) {
+    if (Build.VERSION.SDK_INT >= 28) {
+      ImageDecoder.Source source = ImageDecoder.createSource(buffer);
+      try {
+        return ImageDecoder.decodeBitmap(
+            source,
+            (decoder, info, src) -> {
+              // i.e. ARGB_8888
+              decoder.setTargetColorSpace(ColorSpace.get(ColorSpace.Named.SRGB));
+              decoder.setAllocator(ImageDecoder.ALLOCATOR_SOFTWARE);
```

Comment:
Should have a TODO here to use the hardware allocator when available

Reply:
Done.